core/src/main/java/org/apache/brooklyn/core/mgmt/persist/DeserializingClassRenamesProvider.java

```diff
@@ -20,65 +20,217 @@
+            cache = null;
+        }
+    }
+
+    public interface Loader {
+        public Map<String, String> load();
     }
     
-    private static Map<String, String> loadDeserializingClassRenamesCache() {
-        InputStream resource = new ResourceUtils(DeserializingClassRenamesProvider.class).getResourceFromUrl(DESERIALIZING_CLASS_RENAMES_PROPERTIES_PATH);
-        if (resource != null) {
+    /**
+     * Loads the class-renames from the OSGi configuration file: {@code org.apache.brooklyn.classrename.cfg}.
+     * 
+     * Only public for OSGi instantiation - treat as an internal class, which may change in 
+     * future releases.
+     * 
+     * See http://stackoverflow.com/questions/18844987/creating-a-blueprint-bean-from-an-inner-class:
+     * we unfortunately need to include {@code !org.apache.brooklyn.core.mgmt.persist.DeserializingClassRenamesProvider}
+     * in the Import-Package, as the mvn plugin gets confused due to the use of this inner class
+     * within the blueprint.xml.
+     * 
+     * @see {@link DeserializingClassRenamesProvider#KARAF_DESERIALIZING_CLASS_RENAMES_PROPERTIES}
+     */
+    public static class OsgiConfigLoader implements Loader {
+        private ConfigurationAdmin configAdmin;
+
+        public OsgiConfigLoader() {
+            LOG.trace("OsgiConfigLoader instance created");
+        }
+
+        // For injection as OSGi bean
+        public void setConfigAdmin(ConfigurationAdmin configAdmin) {
+            this.configAdmin = configAdmin;
+        }
+
+        // Called by OSGi
+        public void init() {
+            LOG.trace("DeserializingClassRenamesProvider.OsgiConfigLoader.init: registering loader");
+            DeserializingClassRenamesProvider.loaders.add(this);
+            DeserializingClassRenamesProvider.reset();
+        }
+
+        // Called by OSGi
+        public void destroy() {
+            LOG.trace("DeserializingClassRenamesProvider.OsgiConfigLoader.destroy: unregistering loader");
+            boolean removed = DeserializingClassRenamesProvider.loaders.remove(this);
+            if (removed) {
+                DeserializingClassRenamesProvider.reset();
+            }
+        }
+
+        // Called by OSGi when configuration changes
+        public void updateProperties(Map properties) {
+            LOG.debug("DeserializingClassRenamesProvider.OsgiConfigLoader.updateProperties: clearing cache, so class-renames will be reloaded");
+            DeserializingClassRenamesProvider.reset();
+        }
+
+        @Override
+        public Map<String, String> load() {
+            if (configAdmin == null) {
+                LOG.warn("No OSGi configuration-admin available - cannot load {}.cfg", KARAF_DESERIALIZING_CLASS_RENAMES_PROPERTIES);
+                return ImmutableMap.of();
+            }
+            
+            String filter = '(' + Constants.SERVICE_PID + '=' + KARAF_DESERIALIZING_CLASS_RENAMES_PROPERTIES + ')';
```

Comment:
@geomacy I believe `updateCallback` is not called on the initial creation of the bean, only after configuration changes subsequently. This will cause the properties to be loaded only after someone goes and edits the file post-boot.